PR: Auto-install spyder-kernels into external environment

mrclary opened the pull request:
<!--- Make sure to read the Contributing Guidelines:                   --->
<!--- https://github.com/spyder-ide/spyder/blob/master/CONTRIBUTING.md --->
<!--- and follow PEP 8, PEP 257 and Spyder's code style:               --->
<!--- https://github.com/spyder-ide/spyder/wiki/Dev:-Coding-Style      --->

## Description of Changes

* When the selected Python interpreter environment does not have a compatible `spyder-kernels` version installed, offer to install it.
* If the user rejects the installation, then the original error remains displayed in the console.
* If the user chooses to allow Spyder to install `spyder-kernels`, the user is informed of the changes that will be made to the environment (via `--dry-run`) in a dialog window.
* If an error occurs (including failure to solve the environment), the error is displayed in the console using the info widget.
* If the user chooses not to proceed with the installation, then the original error remains displayed in the console.
* If the user chooses to proceed with the installation, then `spyder-kernels` is installed into the environment and the progress is displayed in the dialog window.
* The user can cancel the operation at any time.
* Upon successful completion the kernel is automatically started and connected.

### Issue(s) Resolved

<!--- List the issue(s) below, in the form "Fixes #1234"; one per line --->

Fixes #22855

![install-spyder-kernels](https://github.com/user-attachments/assets/e87c0abf-d919-4091-abda-29ab268c3ecd)

mrclary committed:
Create function for obtaining channel/url from which a package is installed.

mrclary committed:
Create SpyderKernelVersionError exception and raise this error if an environment does not have the appropriate spyder-kernels version.

mrclary committed:
Create SpyderKernelInstallWidget for installing spyder-kernels

mrclary committed:
Implement SpyderKernelInstallWidget

mrclary committed:
Revise instructions for non-stable spyder-kernels versions

mrclary committed:
Fix test_show_spyder_kernels_error_on_restart

mrclary committed:
Fix "TypeError: unsupported operand type(s) for |: 'type' and 'NoneType'" for pip Python 3.9.

mrclary committed:
Fix test_python_interpreter.
kernel_spec.argv will no longer reset the configuration to the default interpreter if spyder-kernels version is not satisfied in the custom interpreter. Rather a SpyderKernelVersionError is raised to branch the code execution to potentially install spyder-kernels.

CAM-Gerlach commented:
UI wise, the message looks rather different from the other messages we show in the console, in terms of layout, background, text, etc, where it looks more like regular console content than the other pane messages we show (errors, empty pane messages, help pane empty message, etc). Maybe for a followup PR for Carlos/Andres to work on, but perhaps some low hanging fruit fixes would be practical in this PR, e.g. centering the text and perhaps adding a border like the other error/informational messages.

Also, to be clearer that clicking the "OK" button will actually execute the install on their environment rather than merely just accepting that the environment is outdated, you might want to rename it "Install". Or, perhaps to further clarify what the other option is and that it isn't just giving up, the options could be "Install automatically" and "Install manually"?

> I assume that you mean multiple consoles connected to the same kernel or using the same environment. I'm not sure what to do about that. Since this is being handled by the client, multiple clients using the same environment could independently offer to install spyder-kernels. Maybe some kind of check could be run through the main widget that would prevent multiple offerings on the same environment?

It does seem to me that this scenario isn't that likely in practice for actual users, since they are going to get the message on the first console they start in that environment and presumably going to decide to either accept and let it finish or dismiss it before opening another one in the same environment, expecting different results. However, it does seem worth at least making sure that unexpected behavior doesn't result if the user requests multiple installations, e.g. by disabling the "OK" button or not offering the installation if the user opens another console, though it would be better to show a dedicated message that an install is in progress and ideally auto-restart the console when it completes.

mrclary committed:
Ensure that installwidget is below infowidget.
Adding infowidget to layout in refresh_container is redundant and results in infowidget placed below installwidget.

mrclary committed:
Use QMessageBox to prompt user to install spyder-kernels.

mrclary committed:
Disable action to create new console in an environment that is installing spyder-kernels.
Note that setIconVisibleInMenu(False) at SpyderAction instantiation results in the environment menu item not reflecting its enabled state in the menubar. Removing this, or setting it to True (QAction defaults to True), produces the correct behavior. Also note that menubar items still do not show the icon for macOS (as desired) because setIconVisibleInMenu is called in set_menu_icons.

mrclary committed:
Revert "Fix test_show_spyder_kernels_error_on_restart"

This reverts commit 2ac8acd0f6c7d3425ebc63e1434bed08df00449e.

mrclary committed:
Revert "Fix test_python_interpreter.
kernel_spec.argv will no longer reset the configuration to the default interpreter if spyder-kernels version is not satisfied in the custom interpreter. Rather a SpyderKernelVersionError is raised to branch the code execution to potentially install spyder-kernels."

This reverts commit dde7a42b2baf6547172539f381297f64d3d77114.

mrclary commented:
I don't know why the tests are taking so long and result in being cancelled. The failing tests pass locally on my Windows machine 🤷🏼 .

mrclary committed:
Add mamba option to find_conda.
Additionally add the custom conda executable from preferences.

mrclary committed:
Use mamba and fall back to conda.
Use recommended install commands.

mrclary committed:
Client does not need the install message box to be an attribute

mrclary committed:
Ensure original error page is restored if user cancels during install

mrclary committed:
squash messagebox

mrclary committed:
On closing the client, cancel the install widget process if it is running.
It is not possible to close the client while the install message box is active because it is modal; so it is not necessary to reject it here.

mrclary committed:
Assign python executable path attribute on show_install_mbox.

mrclary commented:
Yes, sorry, I haven't pushed up the changes yet. Will do right away.

mrclary committed:
Inform user of what will be changed before proceeding to the install.

mrclary committed:
Reorganize methods

mrclary committed:
Disable environment menu item on dry run

mrclary committed:
Use a common base class with QPlainTextEdit and QProcess

CAM-Gerlach reviewed (commented):
Hey Ryan, some comments on the UI/UX:

- Can you make the modal dialog box asking whether to proceed with the install maybe 50% wider so it doesn't cut off the env path and cause distracting line breaks?
- We discussed just keeping the installation output in the same dialog rather than showing a loading spinner in the kernel, I don't actually have a strong opinion on that and am fine with whatever you think is best
- We discussed making it more obvious that a user needs to make a choice to proceed (it wasn't too confusing to me but @jitseniesen said it might be to others)
- To improve performance, it would make sense to re-using the existing solve result rather than resolving again. @mrclary will explore potential proper ways to do that as well as the backup plan of just using the interactive UI
  - For that we will need to check with @ccordoba12 or @dalthviz that Conda's UI text is always fixed and is not localized to determine when we have reached the interactive prompt, unless there is a better way to do it.
- The CI failures I saw _should_ be fixable in a rebase on the latest `master`

mrclary committed:
Some general cleanup

mrclary committed:
Explicitly include python's install channel for conda environments

mrclary committed:
Perform both dry-run and install from the dialog

mrclary commented:
Catching the user prompt from the process appears to work well. Although there are a few possible issues:
* As we already mentioned, there may be some localization to the prompt. It doesn't appear to have any localization when I looked at conda's and mamba's source code, but I could be wrong.
* Pip does not provide a prompt when installing, so there would be a lack of parity here. The only option is to either retain a dry run for pip before installing, or to just proceed without prompting the user. Pip's dry run output is _very_ obtuse, so I'm not sure how valuable it will be to any but the most astute users.

mrclary commented:
I've updated the screen cast in the OP to reflect the current state of the PR. This is two mamba install processes, both in the dialog window.
